scalableminds/webknossos
# Make NML import more performant and resilient

- Better performance, by keeping trees and groups mutable until put into the store (From ~5s to ~1.5s for this part for a 200,000 node tracing)
- Allow missing tree names and assign a new tree name for missing and empty tree names
- Align tree timestamp assignment to backend NML parser. Replace invalid timestamps with a default timestamp of 0.
- Fixed that validation errors are shown for NML-like files
- Split unconnected trees during NML import and wrap them in a group.

Ideally, I'd like to avoid the redefinition of the `MutableTree` type and, therefore, experimented with recursive read-only instructions for flow, but I couldn't get those to work for the diffable maps :/
(for reference, the most success I had was with the solution outlined in https://github.com/facebook/flow/issues/5844#issuecomment-407169272)

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- [knossos_nmls.zip](https://github.com/scalableminds/webknossos/files/4504148/knossos_nmls.zip)
- Import these knossos NMLs in the tracing view.
  - Unconnected trees should be split
  - There should be no errors during saving
- Try to import an NML where the trees have no name or the nodes have no attributes (except for position and id)
- Check that validation errors are shown for NML-like files, e.g. by removing the id attribute of one of the nodes before upload.

### Issues:
- contributes to #4374 (needs to be done in backend as well)
- fixes #4548 

------
